### PR TITLE
Feature: LendingTerm

### PR DESCRIPTION
**Pull request by eswak**
The LendingTerm is the base of the protocol : it defines loan parameters and allow people to borrow CREDIT, repay loans, call loans, or seize collateral of loans (to be sent to the auction house, a future feature, currently a mock).

Each active LendingTerm will be a gauge in GuildToken, which will set the debt ceiling (max number of CREDIT mintable) for each LendingTerm.

Unit tests : 
image

Coverage : 
image

### CODE REVIEW

**Comment by OneTrueKirk** [comment]
I've been reflecting on whether or not a `LendingTerm` should have an expiration date (beyond which it must be re approved with a new veto period before being used to originate new loans) or not. In my original written designs they did, and this was the intended solution to handle stale prices (for example, we might have an hourly quote for ETH, and a weekly quote for OUSG).

After we spoke @eswak , you were in favor of an optimistic pattern where terms stay good once approved. I agree after thinking it over. With a veto or blacklist mechanism, there's no need to unnecessarily re-approve sound terms, while terms can be marked stale immediately as required by an honest minority.

It feels like simplest way to do this would be a bool in the gauge where a certain amount of `GUILD` can hit the off switch as needed, and then a new vote is needed to turn it back on with the usual veto delay for onboarding a new term.

File: src/loan/LendingTerm.sol [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+        require(GuildToken(_guildToken).isGauge(address(this)), "LendingTerm: terms unavailable");
+
+        // check the debt ceiling & hardcap
+        uint256 _totalBorrows = totalBorrows();
+        require(_totalBorrows + borrowAmount <= hardCap, "LendingTerm: hardcap reached");
+        uint256 _totalSupply = ERC20(creditToken).totalSupply();
+        if (_totalSupply != 0) {
+            uint256 debtCeiling = GuildToken(_guildToken).calculateGaugeAllocation(address(this), _totalSupply + borrowAmount);
+            require(_totalBorrows + borrowAmount <= debtCeiling, "LendingTerm: debt ceiling reached");
+        }
```
Comment (eswak) [inline src/loan/LendingTerm.sol:223]: what happens if the system is perfectly balanced ? say 50%/50% votes on 2 terms and both terms have the same borrow amount

File: src/loan/LendingTerm.sol [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+        require(_originationTime < block.timestamp, "LendingTerm: loan opened in same block");
+
+        // check that the loan is not already called
+        require(loan.callTime == 0, "LendingTerm: loan called");
+
+        // check that the loan is not already closed
+        require(loan.closeTime == 0, "LendingTerm: loan closed");
+    
+        // calculate the fee, pull it from caller, and burn the CREDIT
+        uint256 loanCallFee = loan.borrowAmount * callFee / 1e18;
```
Comment (eswak) [inline src/loan/LendingTerm.sol:303]: here use the public getter

File: src/loan/LendingTerm.sol [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+import {CoreRoles} from "@src/core/CoreRoles.sol";
+import {GuildToken} from "@src/tokens/GuildToken.sol";
+import {CreditToken} from "@src/tokens/CreditToken.sol";
+import {AuctionHouse} from "@src/loan/AuctionHouse.sol";
+import {RateLimitedCreditMinter} from "@src/rate-limits/RateLimitedCreditMinter.sol";
+
+// TODO:
+// - Add a debtDiscountRate in CREDIT to mark down when bad debt occur, and use it to correct all
+//   loan debt amounts in LendingTerm.
+// - Consider if only borrow() should be pausable, or also repay(), call(), and seize()
```
Comment (eswak) [inline src/loan/LendingTerm.sol:16]: consider making all of them pausable separately

File: src/loan/LendingTerm.sol [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+        ERC20(_creditToken).transferFrom(msg.sender, address(this), loanDebt);
+        RateLimitedCreditMinter(creditMinter).replenishBuffer(loanDebt);
+        CreditToken(_creditToken).burn(loanDebt);
+
+        // close the loan
+        loan.closeTime = block.timestamp;
+
+        // update total borrows
+        totalBorrowsStored = totalBorrows() - loanDebt;
+        totalBorrowsLastUpdate = block.timestamp;
```
Comment (eswak) [inline src/loan/LendingTerm.sol:278]: what happens here if the loan has been called ? loanDebt is discounted by the callFee so totalBorrowsStored might not be correct

**Pull request merged**